android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/PackageActivity.java

```diff
@@ -0,0 +1,230 @@
+      }
+
+      @Override
+      public boolean shouldOverrideUrlLoading(WebView view, String url) {
+        if (!url.toLowerCase().equals("about:blank"))
+          view.loadUrl(url);
+
+        return true;
+      }
+
+      @Override
+      public void onPageStarted(WebView view, String url, Bitmap favicon) {
+      }
+
+      @Override
+      public void onPageFinished(WebView view, String url) {
+      }
+    });
+
+    // Determine if KMP contains welcome.htm (case-insensitive) to display
+    FileFilter welcomeFilter = new FileFilter() {
+      @Override
+      public boolean accept(File pathname) {
+        if (pathname.isFile() && pathname.getName().equalsIgnoreCase("welcome.htm")) {
+          return true;
+        }
+        return false;
+      }
+    };
+
+    File[] files = tempPackagePath.listFiles(welcomeFilter);
+    if (files.length > 0 && files[0].exists() && files[0].length() > 0) {
+      webView.loadUrl("file:///" + files[0].getAbsolutePath());
+    } else {
+      String htmlString = "<div id='welcome'><H1>Package: " + pkgId + "</H1></div>";
+      webView.loadData(htmlString, "text/html; charset=utf-8", "UTF-8");
+    }
+
+    installButton.setOnClickListener(new OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        try {
+          // Remove currently installed package and then install
+          File currentPackage = new File(context.getDir("data", MODE_PRIVATE).toString() +
+            File.separator + KMManager.KMDefault_AssetPackages + File.separator + pkgId);
+          if (currentPackage.exists()) {
+            FileUtils.deleteDirectory(currentPackage);
+          }
+          installedPackageKeyboards = PackageProcessor.processKMP(kmpFile);
+          // Do the notifications!
+          boolean success = installedPackageKeyboards.size() != 0;
+          if(success) {
+            notifyPackageInstallListeners(KeyboardEventHandler.EventType.PACKAGE_INSTALLED, installedPackageKeyboards, 1);
+            if(installedPackageKeyboards != null) {
+              notifyPackageInstallListeners(KeyboardEventHandler.EventType.PACKAGE_INSTALLED, installedPackageKeyboards, 1);
+            }
+          } else {
+            Toast.makeText(context, "No new keyboards installed", Toast.LENGTH_SHORT).show();
+            Log.d("PackageActivity", "Package install returned no updated keyboards!");
+          }
+
+        } catch (Exception e) {
+          Log.e("PackageActivity", "Error " + e);
+        } finally {
+          cleanup();
+        }
+      }
+    });
+
+    cancelButton.setOnClickListener(new OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        cleanup();
+      }
+    });
+
+
+  }
+
+  private void cleanup() {
+    try {
+      if (kmpFile.exists()) {
+        kmpFile.delete();
+      }
+      if (tempPackagePath.exists()) {
+        FileUtils.deleteDirectory(tempPackagePath);
+      }
+    } catch (Exception e) {
+
```

Comment:
Do we want to log exceptions here? Can probably help us if we have bugs elsewhere that surface in cleanup.

Reply:
👍